### PR TITLE
feat: Implement MPT changes

### PR DESCRIPTION
**Pull request by shawnxie999**
Spec is here https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-multi-purpose-tokens

Builds rippled on https://github.com/gregtatcam/rippled/tree/mpt-direct-pay

- refactor `Issue` and `STAmount` related changes (will need to change if rippled has further changes)
- introduces new table `mp_token_holders`
- ETL that extracts holders into `mp_token_holders` table whenever a new `MPTokenAuthorize` tx is observed
- new `mpt_holders` API
- inject synthetic `mpt_issuance_id` field into `MPTokenIssuanceCreate` transaction metadata for APIs (eg, `tx`, `ledger`, `subscribe`, `account_tx`)
- support new `MPTokenIssuance` and `MPToken` objects types for `ledger_entry` 
- inject synthetic `mpt_issuance_id` field into `MPTokenIssuance` object for `ledger_data` and `account_objects` APIs

### CODE REVIEW

**Comment by godexsoft** [comment]
@shawnxie999 you need to setup `clangd` locally in your VSCode, or at least some modern version of `clang-format` via an extension. Let me know if you need help to set that up 👍 
Another thing you are probably missing is our githook for checking format on commit. Take a look at our [guide](https://github.com/XRPLF/clio/blob/develop/CONTRIBUTING.md#install-git-hooks).

File: src/etl/MPTHelpers.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * @brief Pull MPT data from ledger object via loadInitialLedger.
+ *
+ * @param key The owner key
+ * @param blob Object data as blob
+ * @return The MPTIssuanceID and holder pair as a optional
+ */
+std::optional<std::pair<ripple::uint192, ripple::AccountID>>
+getMPTHolderFromObj(std::string const& key, std::string const& blob);
+
+}  // namespace etl
```
Comment (godexsoft) [inline src/etl/MPTHelpers.hpp]: In your IDE you can setup to add new line at the end of the file automatically. This will make github happy 👍

File: src/data/Types.hpp [hunk 2]
```diff
@@ -169,6 +169,11 @@ struct NFTsAndCursor {
     std::optional<ripple::uint256> cursor;
 };
 
+struct MPTHoldersAndCursor {
```
Comment (godexsoft) [inline src/data/Types.hpp:239]: Doxygen comments missing. 
It would also be great if you want to add one for the above NFT stuff that is missing them too but that's up to you 👍

File: src/data/cassandra/Schema.hpp [hunk 3]
```diff
@@ -268,6 +268,21 @@ class Schema {
             settingsProvider_.get().getTtl()
         ));
 
+        statements.emplace_back(fmt::format(
+            R"(
+           CREATE TABLE IF NOT EXISTS {}
+                  ( 
+                    mpt_id blob,
+                    holder  blob,
```
Comment (godexsoft) [inline src/data/cassandra/Schema.hpp]: Formatting is not the same as with other existing create statements.. it's very minor but will not pass review 🔢

**Review by godexsoft — COMMENTED** [review]
Adding a few nits in advance. This is not a real review, just a few things that are immediately visible by glancing thru the code 🔢

File: src/data/cassandra/Schema.hpp (hunk 3 above)
Comment (shawnxie999) [reply src/data/cassandra/Schema.hpp]: Im already on clang format version 17. And it gives me this format when I run `clang-format -i ` on this file

File: src/data/Types.hpp (hunk 2 above)
Comment (shawnxie999) [reply src/data/Types.hpp:239]: added

File: src/data/cassandra/Schema.hpp (hunk 3 above)
Comment (godexsoft) [reply src/data/cassandra/Schema.hpp]: Clang-format does nothing to user-provided strings. I'm referring to the formatting of the CQL inside the string.

File: src/data/CassandraBackend.hpp [hunk 4]
```diff
@@ -849,6 +894,20 @@ class BasicCassandraBackend : public BackendInterface {
         executor_.write(std::move(statements));
     }
 
+    void
+    writeMPTHolders(std::vector<std::pair<ripple::uint192, ripple::AccountID>>&& data) override
```
Comment (cindyyan317) [inline src/data/CassandraBackend.hpp]: Do we really need r value here? seems no difference if we just pass a const&

Comment (shawnxie999) [reply src/data/CassandraBackend.hpp]: Will change

File: src/data/CassandraBackend.hpp [hunk 5]
```diff
@@ -849,6 +890,20 @@ class BasicCassandraBackend : public BackendInterface {
         executor_.write(std::move(statements));
     }
 
+    void
+    writeMPTHolders(std::vector<std::pair<ripple::uint192, ripple::AccountID>> const& data) override
+    {
+        std::vector<Statement> statements;
+        for (auto const& record : data) {
+            auto const mpt_id = record.first;
```
Comment (cindyyan317) [inline src/data/CassandraBackend.hpp]: we don't use underscores name such variable

File: src/rpc/RPCHelpers.cpp [hunk 6]
```diff
@@ -314,6 +315,65 @@ insertDeliveredAmount(
...
+    std::shared_ptr<ripple::TxMeta const> const& meta
+)
+{
+    if (!canHaveMPTIssuanceID(txn, meta))
+        return false;
+
+    if (auto const amt = getMPTIssuanceID(meta))
+        metaJson[JS(mpt_issuance_id)] = ripple::to_string(*amt);
+
+    return true;
```
Comment (cindyyan317) [inline src/rpc/RPCHelpers.cpp]: Do we want to return true if amt is emtpy ?

Comment (shawnxie999) [reply src/rpc/RPCHelpers.cpp]: good catch. will fix

File: src/rpc/handlers/AccountObjects.cpp [hunk 7]
```diff
@@ -124,7 +124,16 @@ tag_invoke(boost::json::value_from_tag, boost::json::value& jv, AccountObjectsHa
         std::cbegin(output.accountObjects),
         std::cend(output.accountObjects),
         std::back_inserter(objects),
-        [](auto const& sle) { return toJson(sle); }
+        [](auto const& sle) {
+            auto sleJson = toJson(sle);
+
+            // if object type if mpt issuance, inject synthetic mpt id
+            if (sle.getType() == ripple::ltMPTOKEN_ISSUANCE)
+                sleJson[JS(mpt_issuance_id)] =
```
Comment (cindyyan317) [inline src/rpc/handlers/AccountObjects.cpp]: Maybe move it to toJson(ripple::SLE) function if there is no exception 🤔

**Comment by cindyyan317** [comment]
👍  LGTM. Just a few of small things. The CI seems fail to build xrpl , let's fix it and stamp.

**Comment by godexsoft** [comment]
@shawnxie999 please merge develop again when you have a chance. We fixed the detection mechanism for undocumented code and it may show something new

File: src/data/BackendInterface.hpp [hunk 8]
```diff
@@ -337,6 +337,26 @@ class BackendInterface {
         boost::asio::yield_context yield
     ) const = 0;
 
+    /**
+     * @brief Fetches all holders' balances for a MPTIssuanceID
+     *
+     * @param mptID MPTIssuanceID you wish you query.
+     * @param limit Paging limit.
+     * @param cursorIn Optional cursor to allow us to pick up from where we
+     * last left off.
```
Comment (godexsoft) [inline src/data/BackendInterface.hpp]: This probably does not need to be on a separate line?

File: src/data/CassandraBackend.hpp [hunk 9]
```diff
@@ -552,6 +552,47 @@ class BasicCassandraBackend : public BackendInterface {
...
+            mptKeys.push_back(ripple::keylet::mptoken(mptID, holder).key);
+            cursor = holder;
+        }
+
+        auto const mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);
+
+        // need to filter out the objs that don't exist at the ledger seq because these MPT are in no particular time
+        // order. The MPToken at the specified ledger index may not exist.
+        std::vector<Blob> filteredMpt;
+        std::copy_if(mptObjects.begin(), mptObjects.end(), std::back_inserter(filteredMpt), [](Blob mpt) {
```
Comment (godexsoft) [inline src/data/CassandraBackend.hpp]: Probably no need to copy the Blob in the predicate? 
Also, since we are copying anyway, consider using `std::cbegin(mptObjects)` and `cend` (or the member functions). Alternatively, it might be better to `move` them into `filteredMpt` using `std::move_iterator` (see examples in the codebase).

File: src/data/CassandraBackend.hpp [hunk 10]
```diff
@@ -552,6 +552,47 @@ class BasicCassandraBackend : public BackendInterface {
...
+            return mpt.size() != 0;
+        });
+
+        if (mptKeys.size() == limit)
```
Comment (godexsoft) [inline src/data/CassandraBackend.hpp:583]: Can the size ever reach over the limit? Is the code correct in such case?
If it can't (or should not) consider adding an `ASSERT` for it just above.

File: src/data/CassandraBackend.hpp [hunk 11]
```diff
@@ -849,6 +890,20 @@ class BasicCassandraBackend : public BackendInterface {
         executor_.write(std::move(statements));
     }
 
+    void
+    writeMPTHolders(std::vector<std::pair<ripple::uint192, ripple::AccountID>> const& data) override
+    {
+        std::vector<Statement> statements;
+        for (auto const& record : data) {
+            auto const mptId = record.first;
```
Comment (godexsoft) [inline src/data/CassandraBackend.hpp]: Consider using structured bindings in the `for` and then move the copies into `bind`:
```cpp
for (auto [mptId, holder] : data) { ... }
```

File: src/data/cassandra/Schema.hpp [hunk 12]
```diff
@@ -281,6 +281,21 @@ class Schema {
             settingsProvider_.get().getTtl()
         ));
 
+        statements.emplace_back(fmt::format(
+            R"(
+           CREATE TABLE IF NOT EXISTS {}
+                  ( 
+                    mpt_id blob,
+                    holder blob,
+                    PRIMARY KEY (mpt_id, holder)
```
Comment (godexsoft) [inline src/data/cassandra/Schema.hpp]: This one is still misaligned by one space compared to others 🔢

File: src/etl/impl/LedgerLoader.hpp [hunk 13]
```diff
@@ -55,6 +56,7 @@ struct FormattedTransactionsData {
     std::vector<AccountTransactionsData> accountTxData;
     std::vector<NFTTransactionsData> nfTokenTxData;
     std::vector<NFTsData> nfTokensData;
+    std::vector<std::pair<ripple::uint192, ripple::AccountID>> mptHoldersData;
```
Comment (godexsoft) [inline src/etl/impl/LedgerLoader.hpp]: This pair seems to appear quite a bit. Consider making it a struct perhaps.

File: unittests/util/TestObject.cpp [hunk 14]
```diff
@@ -1001,3 +1001,48 @@ CreateLPTCurrency(std::string_view assetCurrency, std::string_view asset2Currenc
...
+    mptoken.setFieldU32(ripple::sfPreviousTxnLgrSeq, 0);
+
+    if (mptAmount)
+        mptoken.setFieldU64(ripple::sfMPTAmount, mptAmount);
+
+    if (lockedAmount)
+        mptoken.setFieldU64(ripple::sfLockedAmount, lockedAmount);
+
+    return mptoken;
+}
```
Comment (godexsoft) [inline unittests/util/TestObject.cpp]: Missing final `\n` here.

File: src/etl/impl/LedgerLoader.hpp [hunk 15]
```diff
@@ -256,6 +262,7 @@ class LedgerLoader {
                 backend_->writeAccountTransactions(std::move(insertTxResult.accountTxData));
                 backend_->writeNFTs(insertTxResult.nfTokensData);
                 backend_->writeNFTTransactions(insertTxResult.nfTokenTxData);
+                backend_->writeMPTHolders(std::move(insertTxResult.mptHoldersData));
```
Comment (godexsoft) [inline src/etl/impl/LedgerLoader.hpp]: This move is not working as you expected it because your `writeMPTHolders` takes a const ref.

File: src/etl/impl/Transformer.hpp [hunk 16]
```diff
@@ -213,6 +213,7 @@ class Transformer {
         backend_->writeAccountTransactions(std::move(insertTxResultOp->accountTxData));
         backend_->writeNFTs(insertTxResultOp->nfTokensData);
         backend_->writeNFTTransactions(insertTxResultOp->nfTokenTxData);
+        backend_->writeMPTHolders(std::move(insertTxResultOp->mptHoldersData));
```
Comment (godexsoft) [inline src/etl/impl/Transformer.hpp]: Same move issue.

File: src/feed/impl/TransactionFeed.cpp [hunk 17]
```diff
@@ -178,6 +178,7 @@ TransactionFeed::pub(
         pubObj[JS(meta)] = rpc::toJson(*meta);
         rpc::insertDeliveredAmount(pubObj[JS(meta)].as_object(), tx, meta, txMeta.date);
         rpc::insertDeliverMaxAlias(pubObj[txKey].as_object(), version);
+        rpc::insertMPTIssuanceID(pubObj[JS(meta)].as_object(), tx, meta);
```
Comment (godexsoft) [inline src/feed/impl/TransactionFeed.cpp:206]: Maybe take a ref to the `meta` object and reuse? Seems like we already have exactly the same above.

File: src/rpc/handlers/LedgerEntry.cpp [hunk 18]
```diff
@@ -139,6 +139,14 @@ LedgerEntryHandler::process(LedgerEntryHandler::Input input, Context const& ctx)
             key = ripple::keylet::xChainCreateAccountClaimID(input.bridge->value(), input.createAccountClaimId.value())
                       .key;
         }
+    } else if (input.mptIssuance) {
+        auto const mptIssuanceID = ripple::uint192{std::string_view(*(input.mptIssuance))};
+        key = ripple::keylet::mptIssuance(mptIssuanceID).key;
+    } else if (input.mptoken) {
+        auto const holder = ripple::parseBase58<ripple::AccountID>(input.mptoken->at(JS(account)).as_string().c_str());
```
Comment (godexsoft) [inline src/rpc/handlers/LedgerEntry.cpp]: Let's try to avoid `.c_str` if possible. 
Use `boost::json::value_to<std::string>()` if you need a string.

Comment (godexsoft) [reply src/rpc/handlers/LedgerEntry.cpp]: And the same for any other places you have `.c_str()` atm. 👍

File: src/rpc/handlers/MPTHolders.cpp [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+{
+    auto const range = sharedPtrBackend_->fetchLedgerRange();
+    auto const lgrInfoOrStatus = getLedgerInfoFromHashOrSeq(
+        *sharedPtrBackend_, ctx.yield, input.ledgerHash, input.ledgerIndex, range->maxSequence
+    );
+    if (auto const status = std::get_if<Status>(&lgrInfoOrStatus))
+        return Error{*status};
+
+    auto const lgrInfo = std::get<LedgerInfo>(lgrInfoOrStatus);
+
```
Comment (godexsoft) [inline src/rpc/handlers/MPTHolders.cpp]: We probably don't need that much space (between each constant). Maybe try to group them somehow.

File: src/rpc/handlers/MPTHolders.cpp [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+    output.ledgerIndex = lgrInfo.seq;
+
+    boost::json::array mpts;
+    for (auto const& mpt : dbResponse.mptokens) {
+        ripple::STLedgerEntry const sle{ripple::SerialIter{mpt.data(), mpt.size()}, keylet::mptIssuance(mptID).key};
+        boost::json::object mptJson;
+
+        mptJson[JS(account)] = toBase58(sle[ripple::sfAccount]);
+        mptJson[JS(flags)] = sle.getFlags();
+        mptJson["mpt_amount"] = toBoostJson(ripple::STUInt64{sle[ripple::sfMPTAmount]}.getJson(JsonOptions::none));
```
Comment (godexsoft) [inline src/rpc/handlers/MPTHolders.cpp]: There are no JS for these atm. i assume. Leaving it here for when they are added - please update here and below 👍

File: unittests/rpc/handlers/AccountObjectsTests.cpp [hunk 21]
```diff
@@ -1629,3 +1629,95 @@ TEST_F(RPCAccountObjectsHandlerTest, LimitMoreThanMax)
...
+
+    auto const handler = AnyHandler{AccountObjectsHandler{backend}};
+    runSpawn([&](auto yield) {
+        auto const output = handler.process(input, Context{yield});
+        ASSERT_TRUE(output);
+        auto const& accountObjects = output->as_object().at("account_objects").as_array();
+        ASSERT_EQ(accountObjects.size(), 1);
+        EXPECT_EQ(accountObjects.front().at("LedgerEntryType").as_string(), "MPToken");
+    });
+}
```
Comment (godexsoft) [inline unittests/rpc/handlers/AccountObjectsTests.cpp]: `\n`

File: unittests/rpc/handlers/LedgerDataTests.cpp [hunk 22]
```diff
@@ -726,3 +726,83 @@ TEST_F(RPCLedgerDataHandlerTest, JsonLimitMoreThanMax)
...
+        EXPECT_TRUE(output->as_object().contains("ledger"));
+        EXPECT_EQ(output->as_object().at("state").as_array().size(), 1);
+        EXPECT_EQ(output->as_object().at("marker").as_string(), INDEX2);
+        EXPECT_EQ(output->as_object().at("ledger_hash").as_string(), LEDGERHASH);
+        EXPECT_EQ(output->as_object().at("ledger_index").as_uint64(), RANGEMAX);
+
+        auto const& objects = output->as_object().at("state").as_array();
+        EXPECT_EQ(objects.front().at("LedgerEntryType").as_string(), "MPToken");
+    });
+}
```
Comment (godexsoft) [inline unittests/rpc/handlers/LedgerDataTests.cpp]: And again :)

**Review by godexsoft — CHANGES_REQUESTED** [review]
Found some things to improve 👍 
Another thing is that your test coverage is slightly negative - please take a look at [codecov for your branch](https://app.codecov.io/gh/XRPLF/clio/pull/1147?dropdown=coverage&src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=XRPLF) to see where it's lacking tests. Anything with red `change%` needs a bit more love.

File: src/data/CassandraBackend.hpp (hunk 9 above)
Comment (shawnxie999) [reply src/data/CassandraBackend.hpp]: i think i'll use std::remove_if instead and mutate mptObjects directly

File: src/feed/impl/TransactionFeed.cpp (hunk 17 above)
Comment (shawnxie999) [reply src/feed/impl/TransactionFeed.cpp:206]: could you elaborate? `meta` is already a shared pointer.

**Comment by shawnxie999** [comment]
@godexsoft as for test coverage, i believe most of the parts that lack coverage are ETL related

File: src/data/CassandraBackend.hpp [hunk 23]
```diff
@@ -578,19 +578,17 @@ class BasicCassandraBackend : public BackendInterface {
...
-        auto const mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);
+        auto mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);
 
-        // need to filter out the objs that don't exist at the ledger seq because these MPT are in no particular time
-        // order. The MPToken at the specified ledger index may not exist.
-        std::vector<Blob> filteredMpt;
-        std::copy_if(mptObjects.begin(), mptObjects.end(), std::back_inserter(filteredMpt), [](Blob mpt) {
-            return mpt.size() != 0;
-        });
+        auto it = std::remove_if(mptObjects.begin(), mptObjects.end(), [](Blob mpt) { return mpt.size() == 0; });
```
Comment (godexsoft) [inline src/data/CassandraBackend.hpp]: I think you can avoid a copy in the predicate still 👍

File: src/feed/impl/TransactionFeed.cpp (hunk 17 above)
Comment (godexsoft) [reply src/feed/impl/TransactionFeed.cpp:206]: Sorry i missed this comment and only seeing it now. I think i meant `pubObj[JS(meta)].as_object()` that you are looking up a second time here

File: src/data/CassandraBackend.hpp [hunk 24]
```diff
@@ -905,6 +944,16 @@ class BasicCassandraBackend : public BackendInterface {
         executor_.write(std::move(statements));
     }
 
+    void
+    writeMPTHolders(std::vector<MPTHolderData> const& data) override
```
Comment (cindyyan317) [inline src/data/CassandraBackend.hpp:948]: An edge case, I don't know if it is possible.  If this MPT have 2 payments in one ledger, it will generate 2 MPTHolderData, right?  The executor_.write will call batch write operation which does not guarantee the execution order.  We need to make sure the final one is written.

File: src/data/CassandraBackend.hpp [hunk 25]
```diff
@@ -547,6 +547,45 @@ class BasicCassandraBackend : public BackendInterface {
...
+        }
+
+        auto mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);
+
+        auto it = std::remove_if(mptObjects.begin(), mptObjects.end(), [](Blob const& mpt) { return mpt.size() == 0; });
+
+        mptObjects.erase(it, mptObjects.end());
+
+        ASSERT(mptKeys.size() <= limit, "Number of keys can't exceed the limit");
+        if (mptKeys.size() == limit)
```
Comment (cindyyan317) [inline src/data/CassandraBackend.hpp:583]: The limit is for the all the holders. After filter out the ones not exist based on the given ledger sequence, it is possible that the size is not limit, but we still need a cursor.

File: src/etl/MPTHelpers.cpp [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+    return {};
+}
+
+std::optional<MPTHolderData>
+getMPTHolderFromTx(ripple::TxMeta const& txMeta, ripple::STTx const& sttx)
+{
+    if (txMeta.getResultTER() != ripple::tesSUCCESS || sttx.getTxnType() != ripple::TxType::ttMPTOKEN_AUTHORIZE)
+        return {};
+
+    return getMPTokenAuthorize(txMeta);
```
Comment (cindyyan317) [inline src/etl/MPTHelpers.cpp:60]: Looks like MPTId and new holder are available in tx info. We don't need to loop the meta data.

File: src/etl/MPTHelpers.cpp [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+            return MPTHolderData{newMPT[ripple::sfMPTokenIssuanceID], newMPT.getAccountID(ripple::sfAccount)};
+        }
+    }
+    return {};
+}
+
+std::optional<MPTHolderData>
+getMPTHolderFromTx(ripple::TxMeta const& txMeta, ripple::STTx const& sttx)
+{
+    if (txMeta.getResultTER() != ripple::tesSUCCESS || sttx.getTxnType() != ripple::TxType::ttMPTOKEN_AUTHORIZE)
```
Comment (cindyyan317) [inline src/etl/MPTHelpers.cpp:57]: Payment is not supported yet?

File: src/rpc/RPCHelpers.cpp [hunk 28]
```diff
@@ -312,6 +314,67 @@ insertDeliveredAmount(
...
+ *
+ * @param txn The transaction
+ * @param meta The metadata
+ * @return true if the transaction can have a mpt_issuance_id
+ */
+static bool
+canHaveMPTIssuanceID(std::shared_ptr<ripple::STTx const> const& txn, std::shared_ptr<ripple::TxMeta const> const& meta)
+{
+    ripple::TxType const tt{txn->getTxnType()};
+    if (tt != ripple::ttMPTOKEN_ISSUANCE_CREATE)
```
Comment (cindyyan317) [inline src/rpc/RPCHelpers.cpp]: We can use txn->getTxnType() directly in if condition?

File: src/rpc/RPCHelpers.cpp [hunk 29]
```diff
@@ -312,6 +314,67 @@ insertDeliveredAmount(
...
+{
+    if (!canHaveMPTIssuanceID(txn, meta))
+        return false;
+
+    if (auto const id = getMPTIssuanceID(meta)) {
+        metaJson[JS(mpt_issuance_id)] = ripple::to_string(*id);
+        return true;
+    }
+
+    return false;
```
Comment (cindyyan317) [inline src/rpc/RPCHelpers.cpp:376]: I think if we reach this line, means we have logic error in **canHaveMPTIssuanceID**. Maybe assert it.
Do we really use the return ?

File: src/data/CassandraBackend.hpp [hunk 30]
```diff
@@ -547,6 +547,45 @@ class BasicCassandraBackend : public BackendInterface {
...
+    MPTHoldersAndCursor
+    fetchMPTHolders(
+        ripple::uint192 const& mptID,
+        std::uint32_t const limit,
+        std::optional<ripple::AccountID> const& cursorIn,
+        std::uint32_t const ledgerSequence,
+        boost::asio::yield_context yield
+    ) const override
+    {
+        auto const holderEntries = executor_.read(
```
Comment (cindyyan317) [inline src/data/CassandraBackend.hpp:559]: Actually , Only this part needs to be virtual. The filter logic is same for different DB.  By reducing the scope of virtual functions, you can unittest more of the logic without real DB.

File: src/etl/MPTHelpers.cpp (hunk 26 above)
Comment (shawnxie999) [reply src/etl/MPTHelpers.cpp:60]: That would not quite work. Because both issuer or holder submit this transaction. When the issuer submits this transaction, it means it could be either authorizing or unauthorizing a holder. When a holder submits this tx, it could either be creating or destroying a MPToken. In our case, we want only want to know the case where _holder creates a new MPToken_, so extracting the MPToken from the metadata is still the most robust approach. Although I know we still could avoid looping through the metadata by writing additional logic, I would prefer using the most robust approach because transactions like MPTokenAuthorize could potential have new features in the future, which could break the ETL logic if it was written in another way.

File: src/etl/MPTHelpers.cpp (hunk 27 above)
Comment (shawnxie999) [reply src/etl/MPTHelpers.cpp:57]: The only transaction type we need to is extract is MPTokenAuthorize because it is the only one that can create a new MPToken object on the ledger. Payment tx does support MPT transfer between accounts, however that has nothing to do with the mpt_holders API

File: src/data/CassandraBackend.hpp (hunk 30 above)
Comment (shawnxie999) [reply src/data/CassandraBackend.hpp:559]: could you clarify what you mean? inside this function, there is another call to the database     `auto mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);`

Comment (cindyyan317) [reply src/data/CassandraBackend.hpp:559]: If we create a function in CassandraBackend.hpp to just select MPT holder, then all the filter logic can move to BackendInterface. If we want to add a new database, you don't need to copy all the logic to new database implementation. How to filter the valid holder is database unrelated. Instead, new database only needs to provide how to select MPT holder.

Comment (shawnxie999) [reply src/data/CassandraBackend.hpp:559]: But aren't existing functions in CassandraBackend.hpp also do something similar? BTW `fetchMPTHolders` doesn't only have filter logic, it does make another call to the `objects` table  on line 576:

`auto mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);`

Comment (cindyyan317) [reply src/data/CassandraBackend.hpp:559]: I understood the `fetchMPTHolders` does not only have filter logic. So I suggested that we can add the pure virtual function `doFetchMPTHolders` in BackendInterface, which only obtains the data from `selectMPTHolders` table. Then the filer logic can move to BackendInterface normal function. 

override doFetchfetchMPTHolders in CassandraBackend.hpp

```
    std::pair<std::vector<ripple::uint256>, std::optional<ripple::AccountID>>
    doFetchfetchMPTHolders(
        ripple::uint192 const& mptID,
        std::uint32_t const limit,
        std::optional<ripple::AccountID> const& cursorIn,
        boost::asio::yield_context yield
    ) const override
    {
        auto const holderEntries = executor_.read(
            yield, schema_->selectMPTHolders, mptID, cursorIn.value_or(ripple::AccountID(0)), Limit{limit}
        );

        auto const& holderResults = holderEntries.value();
        if (not holderResults.hasRows()) {
            LOG(log_.debug()) << "No rows returned";
            return {};
        }

        std::vector<ripple::uint256> mptKeys;
        std::optional<ripple::AccountID> cursor;
        for (auto const [holder] : extract<ripple::AccountID>(holderResults)) {
            mptKeys.push_back(ripple::keylet::mptoken(mptID, holder).key);
            cursor = holder;
        }
        return {mptKeys, cursor};
    }
```

Then your `fetchMPTHolders` will be like:
```
MPTHoldersAndCursor
BackendInterface::fetchMPTHolders(
    ripple::uint192 const& mptID,
    std::uint32_t const limit,
    std::optional<ripple::AccountID> const& cursorIn,
    std::uint32_t const ledgerSequence,
    boost::asio::yield_context yield
) const
{
    auto [mptKeys, cursor] = doFetchfetchMPTHolders(mptID, limit, cursorIn, yield);

    auto mptObjects = doFetchLedgerObjects(mptKeys, ledgerSequence, yield);

    auto it = std::remove_if(mptObjects.begin(), mptObjects.end(), [](Blob const& mpt) { return mpt.empty(); });

    mptObjects.erase(it, mptObjects.end());

    ASSERT(mptKeys.size() <= limit, "Number of keys can't exceed the limit");
    if (mptKeys.size() == limit)
        return {mptObjects, cursor};

    return {mptObjects, {}};
}
```


I admit that current CassandraBackend.hpp has similar implementation which is not correct. We will refactor them eventually. Feel free to leave it. We can refactor it in future.

**Review by cindyyan317 — DISMISSED** [review]
💯  All comments are addressed. Looks like ready to go. 👍

File: cmake/deps/libxrpl.cmake [hunk 31]
```diff
@@ -1 +1 @@
-find_package(xrpl REQUIRED CONFIG)
+find_package(xrpl-mpt REQUIRED CONFIG)
```
Comment (kuznetsss) [inline cmake/deps/libxrpl.cmake]: I think this should be reverted before merging

Comment (shawnxie999) [reply cmake/deps/libxrpl.cmake]: Yes these will be changed when it's ready to merge. We'd need wait for the rippled one to merge first

**Comment by shawnxie999** [comment]
updated to rippled v2.3.0 in commit  https://github.com/XRPLF/clio/pull/1147/commits/4863a9fc1b873d1855488a46d0b7ffcd2d92701f

@cindyyan317 @godexsoft current build passes. i think would be ready to merge unless you guys have more feedbacks

File: src/feed/impl/TransactionFeed.cpp (hunk 17 above)
Comment (godexsoft) [reply src/feed/impl/TransactionFeed.cpp:206]: It's minor so let's ignore it.

**Review by godexsoft — APPROVED** [review]
I have not spent too much time reviewing this PR but @cindyyan317 did. Once the new ETL is developed, we will have to move the logic into Extensions anyway and we can improve any nits then. Good job on the implementation @shawnxie999 👍

**Comment by godexsoft** [comment]
@shawnxie999 can you squash the above commits and make sure the resulting one is signed by you? then you can force-push the branch.
